In [1]:
import pandas as pd
import numpy as np
from functools import reduce
import matplotlib.pyplot as plt 
%matplotlib inline
import seaborn as sns

# 1. Import the necessary libraries and read the provided CSVs as a data frame and perform the below steps

Read the CSV file

In [2]:
missing_values = ["n.a.","NA","n/a", "na","-"," " ]
file_1=pd.read_csv("C://Users//mahes//Downloads//Data Set//Data Set//phone_user_review_file_1.csv",encoding='latin-1',na_values = missing_values)
file_2=pd.read_csv("C://Users//mahes//Downloads//Data Set//Data Set//phone_user_review_file_2.csv",encoding='latin-1',na_values = missing_values)
file_3=pd.read_csv("C://Users//mahes//Downloads//Data Set//Data Set//phone_user_review_file_3.csv",encoding='latin-1',na_values = missing_values)
file_4=pd.read_csv("C://Users//mahes//Downloads//Data Set//Data Set//phone_user_review_file_4.csv",encoding='latin-1',na_values = missing_values)
file_5=pd.read_csv("C://Users//mahes//Downloads//Data Set//Data Set//phone_user_review_file_5.csv",encoding='latin-1',na_values = missing_values)
file_6=pd.read_csv("C://Users//mahes//Downloads//Data Set//Data Set//phone_user_review_file_6.csv",encoding='latin-1',na_values = missing_values)

In [3]:
file_1.head()

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   

                                             extract       author  \
0  As a diehard Samsung fan who has had every Sam...   CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...    james0923   
2  Adequate feel. Nice heft. Processor's still sl...     R. Craig   
3  Never disappointed. One of the reasons I've be...  Buster2020    
4  I've now found that i'm in a group of people t...   S Ate Mine   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8

In [4]:
file_1.shape

(374910, 11)

In [18]:
file_2.head()

phone_url       date lang country  source        domain  \
0  /cellphones/leagoo-lead-7/  4/15/2015   en      us  Amazon    amazon.com   
1  /cellphones/leagoo-lead-7/  5/23/2015   en      gb  Amazon  amazon.co.uk   
2  /cellphones/leagoo-lead-7/  4/27/2015   en      gb  Amazon  amazon.co.uk   
3  /cellphones/leagoo-lead-7/  4/22/2015   en      gb  Amazon  amazon.co.uk   
4  /cellphones/leagoo-lead-7/  4/18/2015   en      gb  Amazon  amazon.co.uk   

   score  score_max                                            extract  \
0    2.0       10.0  The telephone headset is of poor quality , not...   
1   10.0       10.0  This is my first smartphone so I have nothing ...   
2    8.0       10.0  Great phone. Battery life not great but seems ...   
3   10.0       10.0      Best 90 quid I've ever spent on a smart phone   
4   10.0       10.0  I m happy with this phone.it s very good.thx team   

          author                                            product  
0           luis  Leagoo Lead7 5.0 Inch HD JDI LTPS Screen 3G Sm...  
1     Mark Lavin  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
2         tracey  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
3  Reuben Ingram  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...  
4         viorel  Leagoo Lead 7 Lead7 MTK6582 Quad core 1GB RAM ...

In [11]:
file_2.shape

(114925, 11)

In [19]:
file_3.head()

phone_url       date lang country  \
0  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  11/7/2015   pt      br   
1  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/  10/2/2015   pt      br   
2  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt      br   
3  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/2/2015   pt      br   
4  /cellphones/samsung-galaxy-s-iii-slim-sm-g3812/   9/1/2015   pt      br   

      source            domain  score  score_max  \
0  Submarino  submarino.com.br    6.0       10.0   
1  Submarino  submarino.com.br   10.0       10.0   
2  Submarino  submarino.com.br   10.0       10.0   
3  Submarino  submarino.com.br    8.0       10.0   
4    Colombo    colombo.com.br    8.0       10.0   

                                             extract  \
0  recomendo, eu comprei um, a um ano, e agora co...   
1  Comprei um pouco desconfiada do site e do celu...   
2  Muito bom o produto, obvio que tem versões me...   
3  Unica ressalva fica para a camera que poderia ...   
4  Rapidez e atenção na entrega. O aparelho é ...   

                           author  \
0                herlington tesch   
1             Luisa Silva Marieta   
2                           Cyrus   
3       Marcela Santa Clara Brito   
4  Claudine Maria Kuhn Walendorff   

                                             product  
0  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
1  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
2  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
3  Samsung Smartphone Samsung Galaxy S3 Slim G381...  
4  Smartphone Samsung Galaxy S3 Slim, Dual Chip, ...

In [12]:
file_3.shape

(312961, 11)

In [20]:
file_4.head()


phone_url        date lang country  \
0  /cellphones/samsung-s7262-duos-galaxy-ace/   3/11/2015   en      us   
1  /cellphones/samsung-s7262-duos-galaxy-ace/  17/11/2015   en      in   
2  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
3  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   
4  /cellphones/samsung-s7262-duos-galaxy-ace/  29/10/2015   en      in   

   source      domain  score  score_max  \
0  Amazon  amazon.com    2.0       10.0   
1  Zopper  zopper.com   10.0       10.0   
2  Amazon   amazon.in    4.0       10.0   
3  Amazon   amazon.in    6.0       10.0   
4  Amazon   amazon.in   10.0       10.0   

                                             extract            author  \
0  was not conpatable with my phone as stated. I ...  Frances DeSimone   
1  Decent Functions and Easy to Operate Pros:- Th...     Expert Review   
2  Not Good Phone such price. Hang too much and v...   Amazon Customer   
3                               not bad for features   Amazon Customer   
4                                  Excellent product               NHK   

                                             product  
0  Samsung Galaxy Star Pro DUOS S7262 Unlocked Ce...  
1                Samsung Galaxy Star Pro S7262 Black  
2  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
3  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)  
4  Samsung Galaxy Star Pro GT-S7262 (Midnight Black)

In [13]:
file_4.shape

(98284, 11)

In [21]:
file_5.head()

phone_url       date lang country      source  \
0  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
1  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
2  /cellphones/karbonn-k1616/  7/13/2016   en      in  91 Mobiles   
3  /cellphones/karbonn-k1616/  4/25/2014   en      in     Naaptol   
4  /cellphones/karbonn-k1616/  4/23/2013   en      in     Naaptol   

          domain  score  score_max  \
0  91mobiles.com    2.0       10.0   
1  91mobiles.com    6.0       10.0   
2  91mobiles.com    4.0       10.0   
3    naaptol.com   10.0       10.0   
4    naaptol.com   10.0       10.0   

                                             extract           author  \
0  I bought 1 month before. currently speaker is ...        venkatesh   
1  I just bought one week back, I have Airtel con...           Venkat   
2  one problem in this handset opera is not worki...           krrish   
3  here Karbonn comes up with an another excellen...  BRIJESH CHAUHAN   
4  What a phone, all so on Naaptol my god 23% off...    Suraj CHAUHAN   

                 product  
0          Karbonn K1616  
1          Karbonn K1616  
2          Karbonn K1616  
3  Karbonn K1616 - Black  
4  Karbonn K1616 - Black

In [14]:
file_5.shape

(350216, 11)

In [22]:
file_6.head()

phone_url        date lang country  \
0  /cellphones/samsung-instinct-sph-m800/   9/16/2011   en      us   
1  /cellphones/samsung-instinct-sph-m800/   2/13/2014   en      us   
2  /cellphones/samsung-instinct-sph-m800/  12/30/2011   en      us   
3  /cellphones/samsung-instinct-sph-m800/  10/18/2008   en      us   
4  /cellphones/samsung-instinct-sph-m800/    9/6/2008   en      us   

          source             domain  score  score_max  \
0    Phone Arena     phonearena.com    8.0       10.0   
1         Amazon         amazon.com    6.0       10.0   
2    Phone Scoop     phonescoop.com    9.0       10.0   
3  HandCellPhone  handcellphone.com    4.0       10.0   
4   Reviewed.com       reviewed.com    6.0       10.0   

                                             extract      author  \
0  I've had the phone for awhile and it's a prett...  ajabrams95   
1  to be clear it is not the sellers fault that t...   Stephanie   
2  Well i love this phone. i have had ton of phon...    snickers   
3  I have had my Instinct for several months now ...         A4C   
4  i have had this instinct phone for about two m...   betaBgood   

                     product  
0        Samsung Instinct HD  
1  Samsung SPH M800 Instinct  
2              Instinct M800  
3           Samsung Instinct  
4           Samsung Instinct

In [3]:
file_6.shape

(163837, 11)

A. Merge all the provided CSVs into one dataFrame.

In [5]:
dfs = [file_1,file_2,file_3,file_4,file_5,file_6]

In [6]:
result = pd.concat(dfs)

In [5]:
result.head()

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   

                                             extract       author  \
0  As a diehard Samsung fan who has had every Sam...   CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...    james0923   
2  Adequate feel. Nice heft. Processor's still sl...     R. Craig   
3  Never disappointed. One of the reasons I've be...  Buster2020    
4  I've now found that i'm in a group of people t...   S Ate Mine   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8

In [6]:
result.shape

(1415133, 11)

B. Explore, understand the Data and share at least 2 observations. 

In [5]:
result.describe()

score  score_max
count  1.351644e+06  1351644.0
mean   8.007060e+00       10.0
std    2.616121e+00        0.0
min    2.000000e-01       10.0
25%    7.200000e+00       10.0
50%    9.200000e+00       10.0
75%    1.000000e+01       10.0
max    1.000000e+01       10.0

From the above 5 point summary the mean value for score is 8.007060e
Maximun value is 1.0000000e

In [6]:
result.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1415133 entries, 0 to 163836
Data columns (total 11 columns):
 #   Column     Non-Null Count    Dtype  
---  ------     --------------    -----  
 0   phone_url  1415133 non-null  object 
 1   date       1415133 non-null  object 
 2   lang       1415133 non-null  object 
 3   country    1415133 non-null  object 
 4   source     1415133 non-null  object 
 5   domain     1415133 non-null  object 
 6   score      1351644 non-null  float64
 7   score_max  1351644 non-null  float64
 8   extract    1395730 non-null  object 
 9   author     1351896 non-null  object 
 10  product    1415132 non-null  object 
dtypes: float64(2), object(9)
memory usage: 129.6+ MB


The data type for different feature is given above

In [7]:
result.isnull().sum()

phone_url        0
date             0
lang             0
country          0
source           0
domain           0
score        63489
score_max    63489
extract      19403
author       63237
product          1
dtype: int64

In [8]:
result.dropna(inplace=True)

In [9]:
result.isnull().sum()

phone_url    0
date         0
lang         0
country      0
source       0
domain       0
score        0
score_max    0
extract      0
author       0
product      0
dtype: int64

Check for duplicate values and remove them, if any

In [10]:
result.duplicated().sum()

4466

From the above 4466 duplicate values found

In [11]:
result=result.drop_duplicates()

In [9]:
result.shape

(1271384, 11)

In [10]:
result.head()

phone_url       date lang country            source  \
0  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us  Verizon Wireless   
1  /cellphones/samsung-galaxy-s8/  4/28/2017   en      us       Phone Arena   
2  /cellphones/samsung-galaxy-s8/   5/4/2017   en      us            Amazon   
3  /cellphones/samsung-galaxy-s8/   5/2/2017   en      us           Samsung   
4  /cellphones/samsung-galaxy-s8/  5/11/2017   en      us  Verizon Wireless   

                domain  score  score_max  \
0  verizonwireless.com   10.0       10.0   
1       phonearena.com   10.0       10.0   
2           amazon.com    6.0       10.0   
3          samsung.com    9.2       10.0   
4  verizonwireless.com    4.0       10.0   

                                             extract       author  \
0  As a diehard Samsung fan who has had every Sam...   CarolAnn35   
1  Love the phone. the phone is sleek and smooth ...    james0923   
2  Adequate feel. Nice heft. Processor's still sl...     R. Craig   
3  Never disappointed. One of the reasons I've be...  Buster2020    
4  I've now found that i'm in a group of people t...   S Ate Mine   

                                             product  
0                                  Samsung Galaxy S8  
1                                  Samsung Galaxy S8  
2  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...  
3                      Samsung Galaxy S8 64GB (AT&T)  
4                                  Samsung Galaxy S8

In [11]:
result.nunique()

phone_url       5508
date            7715
lang              20
country           40
source           284
domain           334
score             86
score_max          1
extract      1219892
author        770504
product        54872
dtype: int64

Explore, understand the Data and share at least 2 observations. 

Following are the two observation

Fom the above i have noticed few duplicated values and removed that

Also 5 point summary is given above

Data type for different column is also shown

# Round off scores to the nearest integers

In [12]:
result['score'] =np.round(result['score'],decimals = 1)

In [13]:
result['score']=result['score'].astype(np.int64)

Check for missing values. Impute the missing values, if any.

missing values has been imputed in the above steps

F. Keep only 1 Million data samples. Use random state=612.

In [14]:
result.sample(n = 612)

phone_url       date lang country  \
196430                   /cellphones/huawei-honor-6/  12/5/2014   fr      fr   
294448                       /cellphones/htc-one-sv/  4/29/2013   en      us   
262574            /cellphones/motorola-droid-bionic/  1/15/2012   en      us   
51769               /cellphones/lenovo-vibe-k4-note/  2/14/2016   en      in   
327256                         /cellphones/lg-kp100/  3/12/2009   de      de   
...                                              ...        ...  ...     ...   
45003                    /cellphones/apple-iphone-5/   3/5/2015   en      us   
78496   /cellphones/samsung-galaxy-s6-edge-sm-g925f/   3/2/2016   en      us   
224424                    /cellphones/samsung-s5230/   4/7/2011   ru      ru   
97860                           /cellphones/hp-ipaq/  6/26/2009   es      es   
27356                  /cellphones/huawei-ascend-p7/   8/1/2014   fr      fr   

         source            domain  score  score_max  \
196430   Amazon         amazon.fr      8       10.0   
294448   Amazon        amazon.com      4       10.0   
262574     CNET  reviews.cnet.com      1       10.0   
51769    Amazon         amazon.in      8       10.0   
327256   Amazon         amazon.de     10       10.0   
...         ...               ...    ...        ...   
45003    Amazon        amazon.com      6       10.0   
78496   Samsung       samsung.com      8       10.0   
224424   Yandex  market.yandex.ru      8       10.0   
97860      Ciao           ciao.es      8       10.0   
27356    Amazon         amazon.fr     10       10.0   

                                                  extract           author  \
196430  j' aime : la fluidité du tél, il est hyper r...          wigerst   
294448                    screen was not flush with phone   jason n mccord   
262574  If you depend on your phone for business DO NO...           jotto1   
51769   The big advantage of this mobile is Dolby Atmo...  Amazon Customer   
327256  Ich habe ein günstiges Handy gesucht, das ich...       Manamana78   
...                                                   ...              ...   
45003   it didnt came with a charger and have scratche...          sinthia   
78496   I have been extremely satisfied with the quali...       jjkoregon    
224424  после нокиа немного непрÐ...         ta-di-ra   
97860   Este Hp iPAQ Data Messenger es el mejor pda en...         Mad3line   
27356   D'abord séduit par son esthétique haut de ga...       JCHilarion   

                                                  product  
196430  Honor 6 Smartphone débloqué 4G Cat 6 (Ecran ...  
294448  HTC One SV 4G LTE Prepaid Android Phone (Boost...  
262574    Lenovo Motorola Droid Bionic (Verizon Wireless)  
51769                   Lenovo Vibe K4 Note (Black, 16GB)  
327256                 LG Electronics KP100 Handy schwarz  
...                                                   ...  
45003   Apple iPhone 5 16GB (White) - Sprint (does not...  
78496                Samsung Galaxy S6 edge 64GB (Sprint)  
224424                              Samsung Star GT-S5230  
97860                              HP iPAQ Data Messenger  
27356   Huawei P7 Smartphone débloqué 4G (Ecran: 5 p...  

[612 rows x 11 columns]

G. Drop irrelevant features. Keep features like Author, Product, and Score

In [12]:
result=result[['author','product','score']]

In [16]:
result.head()

author                                            product  score
0   CarolAnn35                                  Samsung Galaxy S8     10
1    james0923                                  Samsung Galaxy S8     10
2     R. Craig  Samsung Galaxy S8 (64GB) G950U 5.8" 4G LTE Unl...      6
3  Buster2020                       Samsung Galaxy S8 64GB (AT&T)      9
4   S Ate Mine                                  Samsung Galaxy S8      4

In [17]:
result['score'].value_counts()

10    617980
8     291874
2     123218
6     110755
4      68243
9      39384
7      10126
5       4444
1       2854
3       2460
0         46
Name: score, dtype: int64

We can also do the above step  by drop() 

# 2. Answer the following questions.

A. Identify the most rated products.

In [16]:
result.groupby('product')['score'].count().sort_values(ascending=False)

product
Lenovo Vibe K4 Note (White,16GB)                                     5223
Lenovo Vibe K4 Note (Black, 16GB)                                    4389
OnePlus 3 (Graphite, 64 GB)                                          4103
OnePlus 3 (Soft Gold, 64 GB)                                         3557
Huawei P8lite zwart / 16 GB                                          2707
                                                                     ... 
Nokia LUMIA 530 BLANC                                                   1
Nokia KIT PIETON ORIGIN STEREO NOKIA WH-102 5800/E52/C5 noir bulk       1
Nokia Illuvial 6700 Rosa                                                1
Nokia Handy Nokia C1-01 Warm Gray Ohne Branding Ohne Simlock            1
摩托罗拉 C168i                                                      1
Name: score, Length: 54872, dtype: int64

The above are few of the most rated product

B. Identify the users with most number of reviews. 

In [17]:
result['author'].value_counts().head(10)

Amazon Customer    76933
Cliente Amazon     19300
e-bit               8410
Client d'Amazon     7716
Amazon Kunde        4746
einer Kundin        2610
Anonymous           2607
einem Kunden        1898
unknown             1725
Anonymous           1460
Name: author, dtype: int64

The above users have most number of reviews

C. Select the data with products having more than 50 ratings and users who have given more than 50 ratings. Report the shape of the final dataset. 

In [18]:
counts1 = pd.value_counts(result['product'])

In [19]:
ratings= result[result['product'].isin(counts1[counts1 > 50].index)]

In [20]:
ratings.head()

author            product  score
0  CarolAnn35  Samsung Galaxy S8     10
1   james0923  Samsung Galaxy S8     10
4  S Ate Mine  Samsung Galaxy S8      4
5       BDB76  Samsung Galaxy S8     10
6    KLC30306  Samsung Galaxy S8      4

In [21]:
count2=pd.value_counts(ratings['author'])

In [22]:
ratings= ratings[ratings['author'].isin(count2[count2 > 50].index)]

In [23]:
ratings.head(10)

author                  product  score
259             schreef        Samsung Galaxy S8     10
271            Егор        Samsung Galaxy S8     10
272    Владимир        Samsung Galaxy S8     10
273  Александр        Samsung Galaxy S8     10
275        Сергей        Samsung Galaxy S8      8
277             Michael  Samsung Galaxy S8 Zwart      8
310               Marco  Samsung Galaxy S8 Zwart     10
317                 Jan  Samsung Galaxy S8 Zwart      8
363                 Kim  Samsung Galaxy S8 Zwart     10
378              Stefan  Samsung Galaxy S8 Zwart      9

3. Build a popularity based model and recommend top 5 mobile phones. [5 Marks]

In [28]:
mean_rating=result.groupby('product')['score'].mean()

In [29]:
mean_rating.head()

product
'Smartphone Meizu Pro 5, 5,7 pouces avec Exynos 7420 Octa 8 Core Processeur. mémoire RAM 4 Go et 64 Go mémoire...            8.0
'Sony Xperia X (F5122) – White – Dual Sim (Google Android 6.0.1, 5 Display, 2 x CORTEX A72 1.8 GHz + 4 x cortex-a53...    10.0
'Sony Xperia X (F5122) – rosa – Dual Sim (Google Android 6.0.1, 5 Display, 2 x CORTEX A72 1.8 GHz + 4 x cortex-a53...     10.0
(7.62 cm (3 )Afficheur/écran, 2 MPixCaméra;blanc)-Smartphone                                                                 6.0
(CUBOT) GT88 5.5" qHD 1.3GHz MTK6572 2-Core Android 4.2.2 3G Phone 8MP CAM 512MB RAM 4GB ROM                                   8.0
Name: score, dtype: float64

In [30]:
result.groupby('product')['score'].mean().sort_values(ascending=False).head(5)

product
摩托罗拉 C168i                                                                                                  10.0
Cubot - Noir GT72 Smartphone Débloqué écran 4.0" MTK6572 Dual Core Téléphone Portable GPS WiFi                 10.0
Cubot - X6 Noir Téléphone Portable 3G/2G Débloqué écran 5.0" MTK6592 Octa Core Smartphone Dual SIM GPS WiFi    10.0
Protector Cristal Templado Acer Liquid Z530                                                                         10.0
Protector Cristal Templado Asus ZenFone Max ZC550KL                                                                 10.0
Name: score, dtype: float64

As per the popularity model the below products are given good rating by the author

1. 摩托罗拉 C168i     
2. Cubot - Noir GT72 Smartphone Débloqué écran 4.0" MTK6572 Dual Core Téléphone Portable GPS WiFi                 
3. Cubot - X6 Noir Téléphone Portable 3G/2G Débloqué écran 5.0" MTK6592 Octa Core Smartphone Dual SIM GPS WiFi    
4. Protector Cristal Templado Acer Liquid Z530                                                                         
5. Protector Cristal Templado Asus ZenFone Max ZC550KL   

Gruping them by product and rating and finally taking mean of that gives you the average rating of the produc. Then i rearranged and out the suggestion of top 5 mobile phone which is shown in the above code



In [31]:
result.groupby('product')['score'].count().sort_values(ascending=False).head(5)

product
Lenovo Vibe K4 Note (White,16GB)     5223
Lenovo Vibe K4 Note (Black, 16GB)    4389
OnePlus 3 (Graphite, 64 GB)          4103
OnePlus 3 (Soft Gold, 64 GB)         3557
Huawei P8lite zwart / 16 GB          2707
Name: score, dtype: int64

However the rating of the products shown above might not be in recommendation list beacuse the author might not given full rating for the above product

# 4. Build a collaborative filtering model using SVD. You can use SVD from surprise or build it from scratch(Note: Incase you’re building it from scratch you can limit your data points to 5000 samples if you face memory issues). Build a collaborative filtering model using kNNWithMeans from surprise. You can try both user-based and item-based model

In [32]:
from surprise import KNNWithMeans
from surprise import accuracy
from surprise.model_selection import train_test_split
from surprise import Dataset,Reader
from surprise import SVD
from surprise import accuracy

In [33]:
reader = Reader(rating_scale=(1, 10))
data = Dataset.load_from_df(ratings[['author', 'product','score']],reader)

In [59]:
from surprise.model_selection import train_test_split
trainset, testset = train_test_split(data, test_size=0.25,random_state=456)

In [60]:
testset

[('einem Kunden',
  'Samsung Galaxy S7 edge Smartphone, 13,9 cm (5,5 Zoll) Display, LTE (4G)',
  10.0),
 ('Amazon Customer', 'Lenovo Vibe K5 (Grey, VoLTE update)', 2.0),
 ('Amazon-Kunde',
  'Palm Germany GmbH Pre Smartphone O2 Branding schwarz',
  10.0),
 ('e-bit', 'Smartphone Samsung Galaxy S4 Mini GT-I9192', 10.0),
 ('Amazon Customer',
  'Asus Zenfone 2 Laser ZE550KL ( White ,2GB Ram, 16GB )',
  8.0),
 ('einem Kunden', 'Apple iPhone 6s 4,7" 128 GB', 10.0),
 ('Amazon Customer', 'OnePlus 3 (Soft Gold, 64 GB)', 8.0),
 ('Ryan', 'HTC Desire Z Sim Free Mobile Phone', 8.0),
 ('Amazon Customer', 'Lenovo Vibe K5 (Grey)', 4.0),
 ('Amazon Customer',
  'Honor 7 Smartphone (13,2 cm (5,2 Zoll) Touchscreen, 16GB interner Speicher, Android OS) grau',
  10.0),
 ('Antonio',
  'Lenovo Motorola Moto G Smartphone, 4,5 pollici display HD, processore Qualcomm, memoria 16GB, MicroSIM, Android 4.3 OS, fotocamera da 5 MP, Nero [Germania]',
  10.0),
 ('Amazon Customer',
  'LG G2 D800 GSM 4G LTE Unlocked Smartp

In [62]:
svd_model = SVD(n_factors=5,biased=False)
svd_model.fit(trainset)

In [63]:
test_pred = svd_model.test(testset)

In [64]:
accuracy.rmse(test_pred)

RMSE: 2.8082


2.8081630748890696

KNN item based model

In [39]:
algo = KNNWithMeans(k=5, sim_options={'user_based': False})

In [40]:
algo.fit(trainset)

Computing the msd similarity matrix...
Done computing similarity matrix.


In [41]:
test_pred1 = algo.test(testset)

In [42]:
print("Item-based Model : Test Set")
accuracy.rmse(test_pred1, verbose=True)

Item-based Model : Test Set
RMSE: 2.9537


2.9537454775891407

User based model

In [43]:
algo_i = KNNWithMeans(k=5, sim_options={'user_based': True})

In [44]:
algo_i.fit(trainset)

Computing the msd similarity matrix...
Done computing similarity matrix.


In [45]:
test_pred2 = algo_i.test(testset)

In [46]:
print("Item-based Model : Test Set")
accuracy.rmse(test_pred2, verbose=True)

Item-based Model : Test Set
RMSE: 2.9626


2.9626195745944903

# 5. Evaluate the collaborative model. Print RMSE value

In [45]:
print("SVD")
accuracy.rmse(test_pred)

SVD
RMSE: 2.7657


2.7656701835351116

In [52]:
print("Item-based Model : Test Set")
accuracy.rmse(test_pred1, verbose=True)

Item-based Model : Test Set
RMSE: 2.9537


2.9537454775891407

In [57]:
print("Item-based Model : Test Set")
accuracy.rmse(test_pred2, verbose=True)

Item-based Model : Test Set
RMSE: 2.9626


2.9626195745944903

From the above i can conclude that RMSE score for SVD is better compared to other two methods

# 6. Predict score (average rating) for test users. 

In [69]:
test_pred

[Prediction(uid='?????????????? ????????????', iid='LG Optimus L5 E612 White', r_ui=10.0, est=8.180852853567751, details={'was_impossible': False}),
 Prediction(uid='einer Kundin', iid='Samsung Galaxy S7 Smartphone, 12,9 cm (5,1 Zoll) Display, LTE (4G)', r_ui=10.0, est=9.330358819772812, details={'was_impossible': False}),
 Prediction(uid='Cliente Amazon', iid='Alcatel Onetouch 2004C - Móvil libre (pantalla 2.4", 0.04 GB RAM, teclas grandes), negro', r_ui=10.0, est=8.622488034112813, details={'was_impossible': False}),
 Prediction(uid='Fabio', iid='Lenovo Motorola Moto G 4G (2 Generazione) Smartphone, Display 5 Pollici, LTE, Fotocamera 8 MP, Memoria 8 GB, Android 5 Lollipop, Nero [Italia]', r_ui=10.0, est=7.4641665294263655, details={'was_impossible': False}),
 Prediction(uid='Amazon Customer', iid='Lenovo Vibe K5 (Gold, VoLTE update)', r_ui=8.0, est=6.4449183413719755, details={'was_impossible': False}),
 Prediction(uid='Amazon Kunde', iid='Huawei Ascend Y530 Smartphone (4,5 Zoll (11

# 8. Try and recommend top 5 products for test users. 

In [47]:
test_pred[20]

Prediction(uid='Silvia', iid='Asus ZenFone 2 Selfie Smartphone, 32 GB, Dual SIM, Aqua Blu [Italia]', r_ui=10.0, est=9.498252186743871, details={'was_impossible': False})

In [49]:
uid = str("Silvia")  
iid = "Asus ZenFone 2 Selfie Smartphone, 32 GB, Dual SIM, Aqua Blu [Italia]"  
# get a prediction for specific users and items.
pred = svd_model.predict(uid, iid, r_ui=10, verbose=True)

user: Silvia     item: Asus ZenFone 2 Selfie Smartphone, 32 GB, Dual SIM, Aqua Blu [Italia] r_ui = 10.00   est = 9.50   {'was_impossible': False}


In [50]:
pred = pd.DataFrame(test_pred)

In [52]:
pred[pred['uid'] =="Silvia" ][['iid', 'r_ui','est']].sort_values(by = 'r_ui',ascending = False).head(10)

iid  r_ui       est
20     Asus ZenFone 2 Selfie Smartphone, 32 GB, Dual ...  10.0  9.498252
15434     Apple iPhone 5C 16GB - sbloccato - blu / ciano  10.0  7.481435
34544                        Apple iPhone 6s 64GB 4G Oro  10.0  9.120998
32497        Nokia Lumia 620 Smartphone, Bianco [Italia]  10.0  8.257816
31356  Lenovo Motorola Moto G 4G (2 Generazione) Smar...  10.0  7.892561
30498          Nokia Lumia 920 Smartphone, Nero [Italia]  10.0  8.911753
29506            Huawei Ascend G6 Smartphone, 4 GB, Nero  10.0  8.356454
26774  Microsoft Lumia 950 Smartphone, 32 GB, Bianco ...  10.0  8.046808
25261    Nokia Lumia 630 Smartphone, 8 GB, Nero [Italia]  10.0  9.399169
22359                          Samsung Galaxy Core Prime  10.0  9.652319

In [65]:
from surprise.model_selection import cross_validate
svd = SVD(verbose=True, n_epochs=10)
cross_validate(svd,data, measures=['RMSE', 'MAE'], cv=3, verbose=True)

Processing epoch 0
Processing epoch 1
Processing epoch 2
Processing epoch 3
Processing epoch 4
Processing epoch 5
Processing epoch 6
Processing epoch 7
Processing epoch 8
Processing epoch 9
Processing epoch 0
Processing epoch 1
Processing epoch 2
Processing epoch 3
Processing epoch 4
Processing epoch 5
Processing epoch 6
Processing epoch 7
Processing epoch 8
Processing epoch 9
Processing epoch 0
Processing epoch 1
Processing epoch 2
Processing epoch 3
Processing epoch 4
Processing epoch 5
Processing epoch 6
Processing epoch 7
Processing epoch 8
Processing epoch 9
Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    2.7012  2.7185  2.7113  2.7103  0.0071  
MAE (testset)     2.1702  2.1802  2.1662  2.1722  0.0059  
Fit time          3.80    6.06    6.66    5.51    1.23    
Test time         0.34    1.10    0.75    0.73    0.31    


{'test_rmse': array([2.70120964, 2.71845113, 2.71132163]),
 'test_mae': array([2.17023591, 2.18017556, 2.16617043]),
 'fit_time': (3.7966458797454834, 6.062127590179443, 6.662686109542847),
 'test_time': (0.3390531539916992, 1.1027114391326904, 0.7484946250915527)}

Before cross validation the RMSE is 2.8

After the cross validation the RMSE is 2.7 

# 10. In what business scenario you should use popularity based Recommendation Systems ?

If the user is new to the environment the popularity based recommended system used. 

In popularity based the most rated product is recommende that is the product with most number of rating from all the user and the avg is considered and it is recomended.

For example, If a person is new to shopping website it will recommend few product based on the other user rating


# 11. In what business scenario you should use CF based Recommendation Systems ?

CF recommendation is recommendation based on the other people with same interest.

For example Person 1 wound have watched movie A,B,C
Person 2 would have watched movie B,C
The system might recommend movie A to person 2 since both of the person have similar interest

# 12. What other possible methods can you think of which can further improve the recommendation for different users ? 

Classification mode:
    This model will use product and user to recommend the product to the particular User

Content based:
    As the name suggest the content of Item will be considered to recommend the product.
    This will makes use of cosine similarity.
    
Hybrid approaches:
    Combining multiple recommendation system techniques is know as Hybrid recommendation System.
    1. Weighted
    2. Switching
    3. Mixed
    4. Feature Combination
    5. Feature Augmentation
    The above are some of the factor which will be considered in Hybrid recommendation system.
    
and the another one is Association rule mining.
Few other like item item , User USer rule and popularity model is already explained above
    
